---
title: "194_BOuru_shotgun_metagenome_analysis"
author: "Shawn Higdon"
date: "2022-10-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load Libraries

```{r, message=FALSE, include=FALSE}

packages <- c("rhdf5", "tidyverse", "ggrepel", "phyloseq", "biomformat",
              "ComplexHeatmap", "RColorBrewer", "viridis", "DESeq2",
              "reshape2", "patchwork", "microbiome", "microbiomeutilities",
              "picante", "ggpubr", "metacoder", "randomcoloR")
lapply(packages, library, character.only = TRUE)
```


## Data Import
### Kraken-Biom
### Sample Map

Kraken Reports were used as input for Bayesian Re-estimation of Taxa Abundance using [**Bracken v2**](https://github.com/jenniferlu717/Bracken). Bracken Report files were converted to biom hdf5 format using the python package [**kraken-biom**](https://github.com/smdabdoub/kraken-biom).

> Use `biomformat` to read in bracken-biom file in hdf5 format

```{r}
# read in hdf5 biom table from kraken-biom output of bracken reports
bracken_biom_file <- read_hdf5_biom("./data/bouru_shotgun_bracken_biome_gtdbr202.hdf5")

# create biom object
bracken_biom <- biom(bracken_biom_file)
bracken_physeq <- import_biom(bracken_biom, parseFunction = parse_taxonomy_greengenes)

# summary of bracken biom
bracken_physeq
head(tax_table(bracken_physeq))

# edit sample names
sample_names(bracken_physeq)
sample_names <- sample_names(bracken_physeq)
sample_names <- str_remove(sample_names, "_bracken")
sample_names(bracken_physeq) <- sample_names
sample_names(bracken_physeq)


# Read in the sample map file (list has strictly static order, numeric ascending)
sample_map <- read.csv("data/sample_map_file.csv", header = T, row.names = 1, stringsAsFactors = T)

# create phyloseq sample map from dataframe
bracken_sam <- sample_data(sample_map, errorIfNULL = TRUE)

bracken_physeq <- merge_phyloseq(bracken_physeq, bracken_sam)

# add tax rank Species2
## Define Ranks to include
#label_ranks <- c("Genus", "Species")
#labels <- apply(tax_table(bracken_physeq)[, label_ranks], 1, paste, sep="", collapse=" ")
## add concatenated labels as a new rank after Species2
#tax_table(bracken_physeq) <- cbind(tax_table(bracken_physeq), Species2=labels)
## remove Kingdom/Domain NA's
tax_table(bracken_physeq) <- tax_table(bracken_physeq)[,1:7]
head(tax_table(bracken_physeq))


ntaxa(bp_milluni)
nsamples(bp_milluni)
sample_names(bp_milluni)
rank_names(bp_milluni)
sample_variables(bp_milluni)
sample_data(bp_milluni)

```

## Evaluate Dataset

### Sample Sums

```{r}
sort(sample_sums(bracken_physeq))
```

__Recommendation: Remove the 5 samples with total read counts below 1 million.__

P1S1 | UruUru_Poopo | San Jose Mine | Mine | sediment
P1WF | UruUru_Poopo | San Jose Mine | acid mine drainage | water filter
P4B1 | UruUru_Poopo | Hot Spring wells | Hot spring | biofilm
P5WF | UruUru_Poopo | Sora Sora River | River | water filter
P4WF | UruUru_Poopo | Hot Spring wells | Hot spring | water filter

__Recommendation: Remove the Biofilm samples from comparative analyses__

### Prune Samples & Taxa

```{r}
# remove the 5 samples with total reads fewer than a million.
bp <- phyloseq::subset_samples(bracken_physeq, phyloseq::sample_sums(bracken_physeq) > 1E6)

# prune taxa
bp2 <- phyloseq::prune_taxa(phyloseq::taxa_sums(bp) > 1000, bp)
bp2
```


## Analyze All Samples

### Data Normalisation and Preprocessing

> The data must be normalized to account for variation among samples in sequencing depth. This is achieved by converting abundances to counts per million.
>
> Additionally, the data must be filtered to prune taxa acting as "noise" with low counts and prevalence overall
>
> CPM normalized counts are then transformed 2 ways:
>
> * Centered-log-ratio transformation
>
> * Relative Abundance
>


```{r Preprocessing bp transform}

# transform to even sampling depth with Phyloseq CPM method
bp_cpm <- transform_sample_counts(bp2, function(x) 1E6 * x/sum(x))

# relative abundance
bp_cpm_relab <- transform_sample_counts(bp_cpm, function(x){ x / sum(x)})

# Centered Log Ratio Transformation (CLR) transform to adjust for sparsity (zeros)
bp_cpm_clr <- transform(bp_cpm, transform = 'clr')

```


### Beta Diversity Tests

```{r}
# Define metadata
metadata <- as(sample_data(bp_cpm_clr), "data.frame")

# calculate microbiome composition distances
clr_dist_mat <- phyloseq::distance(bp_cpm_clr, method = "euclidean")
```


#### Location

##### PERMANOVA
```{r}
#reproducible
set.seed(8888)

# calculate microbiome composition distances
# ADONIS test

## location
permanova_location <- adonis2(phyloseq::distance(bp_cpm_clr, method="euclidean") ~ location,
                     data = metadata, method = "euclidean")
permanova_location
```

##### BETA DISPERSION

```{r}

#Dispersion test and plot
dispr_loc <- vegan::betadisper(clr_dist_mat, phyloseq::sample_data(bp_cpm_clr)$location)
dispr_loc

#Ordination Centroids and Dispersion Labeled: Aitchison Distance
ord_cen <- plot(dispr_loc, main = "", sub = "")
ord_cen

# extract data for ggplot
#ord_sites.long <- BiodiversityR::sites.long(ord_cen)
# boxplot
ord_box_loc <- boxplot(dispr_loc, main = "", xlab = "", xaxt='n')
tick <- seq_along(ord_box_loc$names)
axis(1, at = tick, labels = FALSE)
text(tick, par("usr")[2] - 0, ord_box_loc$names, srt = 45, xpd = TRUE)

# permutation test
permutest(dispr_loc)


```

##### ORDINATION

```{r}
# create sample_id variable
sample_data(bp_cpm_clr)['sample_id'] <- row.names(sample_data(bp_cpm_clr))

# set orination parameters
bp_ord_all_clr <- phyloseq::ordinate(bp_cpm_clr, "RDA")

#Plot scree plot
phyloseq::plot_scree(bp_ord_all_clr) + 
  geom_bar(stat="identity", fill = "blue") +
  labs(x = "\nAxis", y = "Proportion of Variance\n")

#Scale axes and plot ordination
clr1 <- bp_ord_all_clr$CA$eig[1] / sum(bp_ord_all_clr$CA$eig)
clr2 <- bp_ord_all_clr$CA$eig[2] / sum(bp_ord_all_clr$CA$eig)

# plot ordination
bp_ord_all_clr_loc_pca <- phyloseq::plot_ordination(bp_cpm_clr, bp_ord_all_clr, type = "samples", color="location")

# aesthetics of ordination plot
bp_ord_all_clr_loc_pca <- bp_ord_all_clr_loc_pca + 
  geom_point(size = 2) +
  coord_fixed(clr2 / clr1) +
  stat_ellipse(aes(group = location), linetype = 2) +
  labs(color = "Sample Location") +
  geom_text_repel(label = sample_data(bp_cpm_clr)$sample_id, show.legend = FALSE) +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_text(face = "bold", size = 13),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12))

# print ordination plot
pca_allsam_loc <- print(bp_ord_all_clr_loc_pca)

pca_allsam_loc

ggsave("output/plots/pca_gtdb_location.pdf", pca_allsam_loc, dpi = 600)
```


#### Location Type

##### PERMANOVA
```{r}
set.seed(8888)
## location_type
permanova_loc.type <- adonis2(phyloseq::distance(bp_cpm_clr, method="euclidean") ~ location_type,
                     data = metadata, method = "euclidean")
permanova_loc.type
```

##### BETA DISPERSION

```{r}
#Dispersion test and plot
dispr_loc.type <- vegan::betadisper(clr_dist_mat, phyloseq::sample_data(bp_cpm_clr)$location_type)
dispr_loc.type

#Ordination Centroids and Dispersion Labeled: Aitchison Distance
ord_cen_lt <- plot(dispr_loc.type, main = "", sub = "")
ord_cen_lt

# extract data for ggplot
#ord_sites.long <- BiodiversityR::sites.long(ord_cen)
# boxplot
ord_box_lt <- boxplot(dispr_loc.type, main = "", xlab = "")


# permutation test
permutest(dispr_loc.type)
```

##### ORDINATION

```{r}
# plot ordination
bp_ord_all_clr_loc.type_pca <- phyloseq::plot_ordination(bp_cpm_clr, bp_ord_all_clr, type = "samples", color="location_type")

# aesthetics of ordination plot
bp_ord_all_clr_loc.type_pca <- bp_ord_all_clr_loc.type_pca + 
  geom_point(size = 2) +
  coord_fixed(clr2 / clr1) +
  stat_ellipse(aes(group = location_type), linetype = 2) +
  labs(color = "Sample Location Type") +
  geom_text_repel(label = sample_data(bp_cpm_clr)$sample_id, show.legend = FALSE) +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_text(face = "bold", size = 13),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12))

# print ordination plot
pca_allsam_loc.type <- print(bp_ord_all_clr_loc.type_pca)

pca_allsam_loc.type

```



#### Sample Type

##### PERMANOVA
```{r}
set.seed(8888)
## location_type
permanova_sample.type <- adonis2(phyloseq::distance(bp_cpm_clr, method="euclidean") ~ sample_type,
                     data = metadata, method = "euclidean")
permanova_sample.type
```

##### BETA DISPERSION
```{r}
#Dispersion test and plot
dispr_sam.type <- vegan::betadisper(clr_dist_mat, phyloseq::sample_data(bp_cpm_clr)$sample_type)
dispr_sam.type

#Ordination Centroids and Dispersion Labeled: Aitchison Distance
ord_cen_sam <- plot(dispr_sam.type, main = "", sub = "")
ord_cen_sam

# extract data for ggplot
#ord_sites.long <- BiodiversityR::sites.long(ord_cen)
# boxplot
ord_box_sam <- boxplot(dispr_sam.type, main = "", xlab = "")

# permutation test
permutest(dispr_sam.type)
```

> High variance in sediment bacterial microbiota composition


##### ORDINATION

```{r}
# plot ordination
bp_ord_all_clr_sam.type_pca <- phyloseq::plot_ordination(bp_cpm_clr, bp_ord_all_clr, type = "samples", color="sample_type")

# aesthetics of ordination plot
bp_ord_all_clr_sam.type_pca <- bp_ord_all_clr_sam.type_pca + 
  geom_point(size = 2) +
  coord_fixed(clr2 / clr1) +
  stat_ellipse(aes(group = sample_type), linetype = 2) +
  labs(color = "Sample Type") +
  geom_text_repel(label = sample_data(bp_cpm_clr)$sample_id, show.legend = FALSE) +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_text(face = "bold", size = 13),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12))

# print ordination plot
pca_allsam_sam.type <- print(bp_ord_all_clr_sam.type_pca)

pca_allsam_sam.type

```

> Water filter composition appears close to substrata of sediment, but sediment has high variance.

### Alpha Div

#### Obs. Richness vs. Total Reads
```{r}
ggplot(data = data.frame("total_reads" =  phyloseq::sample_sums(bracken_physeq),
                         "observed" = phyloseq::estimate_richness(bracken_physeq, measures = "Observed")[, 1]),
       aes(x = total_reads, y = observed)) +
  geom_point() +
  geom_smooth(method="lm", se = FALSE) +
  labs(x = "\nTotal Reads", y = "Observed Richness\n")
```

> Positive upward trend between Observed Richness and total number of reads for each sample, as expected.

#### Shannon Div

> Boxplot visualizations with wilcox. test for significant differences in Shannon Diversity Index measurements. 

##### Location
```{r}
bp_all_boxplot_loc <- plot_richness(bp_cpm, x = "location", color = "location", measures = c("Shannon"))#, title = "Microbiome Alpha Diversity: Samples by NDFA Group")
bp_all_boxplot_loc <- bp_all_boxplot_loc + geom_boxplot() +
  labs(color="Sample Location") +
  xlab("Location") +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.y = element_text(size = 10))

comps <- make_pairs(sample_data(bp_cpm)$location)
print(comps)

adiv_loc <- bp_all_boxplot_loc + stat_compare_means(
  comparisons = comps,
  label = "p.signif",
  tip.length = 0.05,
  symnum.args = list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
    symbols = c("xxxx", "***", "**", "*", "ns")
  ),
  method = "wilcox.test")

adiv_loc
```


##### Location Type
```{r}
bp_all_boxplot_lt <- plot_richness(bp_cpm, x = "location_type", color = "location_type", measures = c("Shannon"))#, title = "Microbiome Alpha Diversity: Samples by NDFA Group")
bp_all_boxplot_lt <- bp_all_boxplot_lt + geom_boxplot() +
  labs(color="Location Type") +
  xlab("Location Type") +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.y = element_text(size = 10))

comps <- make_pairs(sample_data(bp_cpm)$location_type)
print(comps)

adiv_lt <- bp_all_boxplot_lt + stat_compare_means(
  comparisons = comps,
  label = "p.signif",
  tip.length = 0.05,
  symnum.args = list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
    symbols = c("xxxx", "***", "**", "*", "ns")
  ),
  method = "wilcox.test")

adiv_lt
```

##### Sample Type
```{r}
bp_all_boxplot_sam <- plot_richness(bp_cpm, x = "sample_type", color = "sample_type", measures = c("Shannon"))#, title = "Microbiome Alpha Diversity: Samples by NDFA Group")
bp_all_boxplot_sam <- bp_all_boxplot_sam + geom_boxplot() +
  labs(color="Sample Type") +
  xlab("Sample Type") +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.y = element_text(size = 10))

comps <- make_pairs(sample_data(bp_cpm)$sample_type)
print(comps)

adiv_sam <- bp_all_boxplot_sam + stat_compare_means(
  comparisons = comps,
  label = "p.signif",
  tip.length = 0.05,
  symnum.args = list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
    symbols = c("xxxx", "***", "**", "*", "ns")
  ),
  method = "wilcox.test")

adiv_sam
```

## Taxonomic Composition

```{r}
mycolors <- colorRampPalette(brewer.pal(8, "Paired"))(topN+1)

### DECIDE how many taxa you want to see
topN = 14 # all others will be grouped in "Others"
```


### By Location

#### Top Phyla
```{r}
topPhyla <- bp_cpm %>%  #choose dataset
  psmelt %>% group_by(Phylum) %>% # melt the table and group by tax level
  summarise(Abundance=sum(Abundance)) %>% # find most overall abundant taxa
  arrange(desc(Abundance)) %>% # order them 
  mutate(aggTaxo=as.factor(case_when( # aggTaxo will become the plot legend
    row_number()<=topN~Phylum, row_number()>topN~'Others'))) %>% 
  dplyr::select(-Abundance) %>% #flush abundance value
  head(n=topN+1) # +1 to include the Others section! 

### CREATE an object that will be fed into ggplot
plot_bp_phyla <- bp_cpm %>% # choose dataset
  psmelt %>% # "melt" the phyloseq object into a table
  inner_join(.,topPhyla, by="Phylum") %>% # use topTaxa to join aggTaxo variable
  aggregate(Abundance~location+aggTaxo, # Abundance is aggregated...
            data=., FUN=sum) # ...sums "Others" taxa by specified variables

# Relative abundance counts plot
relab_top_phyla_location <- ggplot(plot_bp_phyla, aes(x=location, y=Abundance, fill=aggTaxo)) +
  geom_bar(stat="identity",position="fill") + # change to "stack" to plot total counts
  labs(title="Top Phyla by location",
       fill="Phylum", x="Location", y="Relative abundance") +
  #facet_wrap(~sample_type) + # RHS of the tilde is columns; add scales="free" to hide samples without counts
  scale_fill_manual(values = mycolors,breaks=c(pull(topPhyla, Phylum),"Others")) +
  theme(axis.line=element_blank(),
        axis.ticks=element_blank(),
        plot.margin = unit(c(1,0.5,1,2), "cm"),
        plot.caption.position="plot",
        axis.title.y.left = element_text(vjust=3),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        axis.title.x = element_blank())

relab_top_phyla_location

ggsave("./output/plots/relab_top_phyla_location_gtdbr202.pdf", relab_top_phyla_location)
```

#### Top Family
```{r}
topFamily <- bp_cpm %>%  #choose dataset
  psmelt %>% group_by(Family) %>% # melt the table and group by tax level
  summarise(Abundance=sum(Abundance)) %>% # find most overall abundant taxa
  arrange(desc(Abundance)) %>% # order them 
  mutate(aggTaxo=as.factor(case_when( # aggTaxo will become the plot legend
    row_number()<=topN~Family, row_number()>topN~'Others'))) %>% 
  dplyr::select(-Abundance) %>% #flush abundance value
  head(n=topN+1) # +1 to include the Others section! 

### CREATE an object that will be fed into ggplot
plot_bp_family <- bp_cpm %>% # choose dataset
  psmelt %>% # "melt" the phyloseq object into a table
  inner_join(.,topFamily, by="Family") %>% # use topTaxa to join aggTaxo variable
  aggregate(Abundance~location+aggTaxo, # Abundance is aggregated...
            data=., FUN=sum) # ...sums "Others" taxa by specified variables

```

```{r}
# Relative abundance counts plot
relab_top_family_location <- ggplot(plot_bp_family, aes(x=location, y=Abundance, fill=aggTaxo)) +
  geom_bar(stat="identity",position="fill") + # change to "stack" to plot total counts
  labs(title="Top Families by location",
       fill="Family", x="Location", y="Relative abundance") +
  #facet_wrap(~sample_type) + # RHS of the tilde is columns; add scales="free" to hide samples without counts
  scale_fill_manual(values = mycolors,breaks=c(pull(topFamily, Family),"Others")) +
  theme(axis.line=element_blank(),
        axis.ticks=element_blank(),
        plot.margin = unit(c(1,0.5,1,2), "cm"),
        plot.caption.position="plot",
        axis.title.y.left = element_text(vjust=3),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        axis.title.x = element_blank())

relab_top_family_location

ggsave("./output/plots/relab_top_family_location_gtdbr202.pdf", relab_top_family_location)

```

#### Top Genera
```{r}
topGenera <- bp_cpm %>%  #choose dataset
  psmelt %>% group_by(Genus) %>% # melt the table and group by tax level
  summarise(Abundance=sum(Abundance)) %>% # find most overall abundant taxa
  arrange(desc(Abundance)) %>% # order them 
  mutate(aggTaxo=as.factor(case_when( # aggTaxo will become the plot legend
    row_number()<=topN~Genus, row_number()>topN~'Others'))) %>% 
  dplyr::select(-Abundance) %>% #flush abundance value
  head(n=topN+1) # +1 to include the Others section! 

### CREATE an object that will be fed into ggplot
plot_bp_genera <- bp_cpm %>% # choose dataset
  psmelt %>% # "melt" the phyloseq object into a table
  inner_join(.,topGenera, by="Genus") %>% # use topTaxa to join aggTaxo variable
  aggregate(Abundance~location+aggTaxo, # Abundance is aggregated...
            data=., FUN=sum) # ...sums "Others" taxa by specified variables

```

```{r}
# Relative abundance counts plot
relab_top_genera_location <- ggplot(plot_bp_genera, aes(x=location, y=Abundance, fill=aggTaxo)) +
  geom_bar(stat="identity",position="fill") + # change to "stack" to plot total counts
  labs(title="Top Genera by location",
       fill="Genus", x="Location", y="Relative abundance") +
  #facet_wrap(~sample_type) + # RHS of the tilde is columns; add scales="free" to hide samples without counts
  scale_fill_manual(values = mycolors,breaks=c(pull(topGenera, Genus),"Others")) +
  theme(axis.line=element_blank(),
        axis.ticks=element_blank(),
        plot.margin = unit(c(1,0.5,1,2), "cm"),
        plot.caption.position="plot",
        axis.title.y.left = element_text(vjust=3),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        axis.title.x = element_blank())


relab_top_genera_location

ggsave("./output/plots/relab_top_genera_location_gtdbr202.pdf", relab_top_genera_location)
```


### By Location Type

#### Top Phyla
```{r}
### CREATE an object that will be fed into ggplot
plot_bp_phyla.lt <- bp_cpm %>% # choose dataset
  psmelt %>% # "melt" the phyloseq object into a table
  inner_join(.,topPhyla, by="Phylum") %>% # use topTaxa to join aggTaxo variable
  aggregate(Abundance~location_type+aggTaxo, # Abundance is aggregated...
            data=., FUN=sum) # ...sums "Others" taxa by specified variables

# Relative abundance counts plot
relab_top_phyla_lt <- ggplot(plot_bp_phyla.lt, aes(x=location_type, y=Abundance, fill=aggTaxo)) +
  geom_bar(stat="identity",position="fill") + # change to "stack" to plot total counts
  labs(title="Top Phyla by location type",
       fill="Phylum", x="Location Type", y="Relative abundance") +
  #facet_wrap(~sample_type) + # RHS of the tilde is columns; add scales="free" to hide samples without counts
  scale_fill_manual(values = mycolors,breaks=c(pull(topPhyla, Phylum),"Others")) +
  theme(axis.line=element_blank(),
        axis.ticks=element_blank(),
        plot.margin = unit(c(1,0.5,1,2), "cm"),
        plot.caption.position="plot",
        axis.title.y.left = element_text(vjust=3),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        axis.title.x = element_blank())

relab_top_phyla_lt

ggsave("./output/plots/relab_top_phyla_location-type_gtdbr202.pdf", relab_top_phyla_lt)
```

#### Top Family
```{r}
### CREATE an object that will be fed into ggplot
plot_bp_family.lt <- bp_cpm %>% # choose dataset
  psmelt %>% # "melt" the phyloseq object into a table
  inner_join(.,topFamily, by="Family") %>% # use topTaxa to join aggTaxo variable
  aggregate(Abundance~location_type+aggTaxo, # Abundance is aggregated...
            data=., FUN=sum) # ...sums "Others" taxa by specified variables

# Relative abundance counts plot
relab_top_family_lt <- ggplot(plot_bp_family.lt, aes(x=location_type, y=Abundance, fill=aggTaxo)) +
  geom_bar(stat="identity",position="fill") + # change to "stack" to plot total counts
  labs(title="Top Families by location type",
       fill="Family", x="Location Type", y="Relative abundance") +
  #facet_wrap(~sample_type) + # RHS of the tilde is columns; add scales="free" to hide samples without counts
  scale_fill_manual(values = mycolors,breaks=c(pull(topFamily, Family),"Others")) +
  theme(axis.line=element_blank(),
        axis.ticks=element_blank(),
        plot.margin = unit(c(1,0.5,1,2), "cm"),
        plot.caption.position="plot",
        axis.title.y.left = element_text(vjust=3),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        axis.title.x = element_blank())

relab_top_family_lt

ggsave("./output/plots/relab_top_family_location-type_gtdbr202.pdf", relab_top_family_lt)
```

#### Top Genera
```{r}
### CREATE an object that will be fed into ggplot
plot_bp_genera.lt <- bp_cpm %>% # choose dataset
  psmelt %>% # "melt" the phyloseq object into a table
  inner_join(.,topGenera, by="Genus") %>% # use topTaxa to join aggTaxo variable
  aggregate(Abundance~location_type+aggTaxo, # Abundance is aggregated...
            data=., FUN=sum) # ...sums "Others" taxa by specified variables

# Relative abundance counts plot
relab_top_genera_lt <- ggplot(plot_bp_genera.lt, aes(x=location_type, y=Abundance, fill=aggTaxo)) +
  geom_bar(stat="identity",position="fill") + # change to "stack" to plot total counts
  labs(title="Top Genera by location type",
       fill="Genus", x="Location Type", y="Relative abundance") +
  #facet_wrap(~sample_type) + # RHS of the tilde is columns; add scales="free" to hide samples without counts
  scale_fill_manual(values = mycolors,breaks=c(pull(topGenera, Genus),"Others")) +
  theme(axis.line=element_blank(),
        axis.ticks=element_blank(),
        plot.margin = unit(c(1,0.5,1,2), "cm"),
        plot.caption.position="plot",
        axis.title.y.left = element_text(vjust=3),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1),
        axis.title.x = element_blank())

relab_top_genera_lt

ggsave("./output/plots/relab_top_genera_location-type_gtdbr202.pdf", relab_top_genera_lt)
```

### Metacoder plots

```{r}
# convert relative abundance
obj <- parse_phyloseq(bp_cpm_relab, class_regex = "(.*)", class_key = "taxon_name")
?parse_phyloseq()
```

```{r}
# inspect taxonomy map object

print(obj) # 7400 taxa


```

```{r}
obj$data$tax_data
```


```{r}
# heat tree

heat_tree(metacoder::filter_taxa(obj, taxon_names == "Bacteria", subtaxa = TRUE),
          node_size = n_obs,
          node_label = taxon_names,
          node_color = n_obs,
          layout = "fruchterman-reingold")
```

###Taxa Summary

> taxa with relative abundances less than 0.01 were removed.

```{r}
bp.tax_abund.summary <- taxa_summary(bp_cpm_relab, "Species") %>%
  arrange(., desc(Mean.Rel.Ab))

# Save

write_csv(bp.tax_abund.summary, "./output/bouru_gtdb_microbiota_taxa_relabund_summary.csv", col_names = T)
```

> Need to revisit this...


```{r}
dom.tax <- dominant_taxa(bp ,level = "Species", group="location_type")
dom.tax$dominant_overview
```



```{r}
test <- as.data.frame(otu_table(bp_cpm))
```










## Milluni Subset Only


```{r}
# subset dataset with 5 low read samples removed
bp_milluni <- subset_samples(bp, project == "Milluni")
bp_milluni
```


```{r}
# prune taxa
bpm2 <- phyloseq::prune_taxa(phyloseq::taxa_sums(bp) > 1000, bp_milluni)
bpm2
```

### Data Normalisation and Preprocessing

> The data must be normalized to account for variation among samples in sequencing depth. This is achieved by converting abundances to counts per million.
>
> Additionally, the data must be filtered to prune taxa acting as "noise" with low counts and prevalence overall
>
> CPM normalized counts are then transformed 2 ways:
>
> * Centered-log-ratio transformation
>
> * Relative Abundance
>


```{r Preprocessing bp transform}

# transform to even sampling depth with Phyloseq CPM method
bpm_cpm <- transform_sample_counts(bpm2, function(x) 1E6 * x/sum(x))

# relative abundance
bpm_cpm_relab <- transform_sample_counts(bpm_cpm, function(x){ x / sum(x)})

# Centered Log Ratio Transformation (CLR) transform to adjust for sparsity (zeros)
bpm_cpm_clr <- transform(bpm_cpm, transform = 'clr')

```


### Beta Diversity Tests

```{r}
# Define metadata
bpm_metadata <- as(sample_data(bpm_cpm_clr), "data.frame")

# calculate microbiome composition distances
bpm_clr_dist_mat <- phyloseq::distance(bpm_cpm_clr, method = "euclidean")
```


#### Location

##### PERMANOVA
```{r}
#reproducible
set.seed(8888)

# calculate microbiome composition distances
# ADONIS test

## location
bpm_permanova_location <- adonis2(phyloseq::distance(bpm_cpm_clr, method="euclidean") ~ location,
                     data = bpm_metadata, method = "euclidean")
bpm_permanova_location
```

##### BETA DISPERSION

```{r}

#Dispersion test and plot
bpm_dispr_loc <- vegan::betadisper(bpm_clr_dist_mat, phyloseq::sample_data(bpm_cpm_clr)$location)
bpm_dispr_loc

#Ordination Centroids and Dispersion Labeled: Aitchison Distance
bpm_ord_cen <- plot(bpm_dispr_loc, main = "", sub = "")
bpm_ord_cen

# extract data for ggplot
#ord_sites.long <- BiodiversityR::sites.long(ord_cen)
# boxplot
bpm_ord_box_loc <- boxplot(bpm_dispr_loc, main = "", xlab = "", xaxt='n')
tick <- seq_along(ord_box_loc$names)
axis(1, at = tick, labels = FALSE)
text(tick, par("usr")[2] - 0, bpm_ord_box_loc$names, srt = 45, xpd = TRUE)

# permutation test
permutest(bpm_dispr_loc)


```

##### ORDINATION

```{r}
# create sample_id variable
sample_data(bpm_cpm_clr)['sample_id'] <- row.names(sample_data(bpm_cpm_clr))

# set orination parameters
bpm_ord_all_clr <- phyloseq::ordinate(bpm_cpm_clr, "RDA")

#Plot scree plot
phyloseq::plot_scree(bpm_ord_all_clr) + 
  geom_bar(stat="identity", fill = "blue") +
  labs(x = "\nAxis", y = "Proportion of Variance\n")

#Scale axes and plot ordination
clr1 <- bpm_ord_all_clr$CA$eig[1] / sum(bpm_ord_all_clr$CA$eig)
clr2 <- bpm_ord_all_clr$CA$eig[2] / sum(bpm_ord_all_clr$CA$eig)

# plot ordination
bpm_ord_all_clr_loc_pca <- phyloseq::plot_ordination(bpm_cpm_clr, bpm_ord_all_clr, type = "samples", color="location")

# aesthetics of ordination plot
bpm_ord_all_clr_loc_pca <- bpm_ord_all_clr_loc_pca + 
  geom_point(size = 2) +
  coord_fixed(clr2 / clr1) +
  stat_ellipse(aes(group = location), linetype = 2) +
  labs(color = "Sample Location") +
  geom_text_repel(label = sample_data(bpm_cpm_clr)$sample_id, show.legend = FALSE) +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_text(face = "bold", size = 13),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12))

# print ordination plot
bpm_pca_allsam_loc <- print(bpm_ord_all_clr_loc_pca)

bpm_pca_allsam_loc

ggsave("output/plots/milluni_pca_gtdb_location.pdf", bpm_pca_allsam_loc, dpi = 600)
```


## UruUru PooPo Subset Only

> GTDB r202 used with Kraken2 / Bracken 2


```{r}
# subset dataset with 5 low read samples removed
bpu <- subset_samples(bp, project == "UruUru_Poopo")
bpu

print(sample_names(bpu))

```


```{r}
# prune taxa
bpu2 <- phyloseq::prune_taxa(phyloseq::taxa_sums(bp) > 1000, bpu)
bpu2

# create 8th tax rank to combine genus and species
label_ranks <- c("Genus", "Species")
labels <- apply(tax_table(bpu2)[, label_ranks], 1, paste, sep="", collapse=" ")
## add concatenated labels as a new rank after Species2
tax_table(bpu2) <- cbind(tax_table(bpu2), Species2=labels)
head(tax_table(bpu2))
```

### Data Normalisation and Preprocessing

> The data must be normalized to account for variation among samples in sequencing depth. This is achieved by converting abundances to counts per million.
>
> Additionally, the data must be filtered to prune taxa acting as "noise" with low counts and prevalence overall
>
> CPM normalized counts are then transformed 2 ways:
>
> * Centered-log-ratio transformation
>
> * Relative Abundance
>


```{r Preprocessing bp transform}

# transform to even sampling depth with Phyloseq CPM method
bpu_cpm <- transform_sample_counts(bpu2, function(x) 1E6 * x/sum(x))

# relative abundance
bpu_cpm_relab <- transform_sample_counts(bpu_cpm, function(x){ x / sum(x)})

# Centered Log Ratio Transformation (CLR) transform to adjust for sparsity (zeros)
bpu_cpm_clr <- transform(bpu_cpm, transform = 'clr')

```


### Beta Diversity Tests

```{r}
# Define metadata
bpu_metadata <- as(sample_data(bpu_cpm_clr), "data.frame")

# calculate microbiome composition distances
bpu_clr_dist_mat <- phyloseq::distance(bpu_cpm_clr, method = "euclidean")
```


#### Location

##### PERMANOVA
```{r}
#reproducible
set.seed(8888)

# calculate microbiome composition distances
# ADONIS test

## location
bpu_permanova_location <- adonis2(phyloseq::distance(bpu_cpm_clr, method="euclidean") ~ location,
                     data = bpu_metadata, method = "euclidean")
bpu_permanova_location
```

##### BETA DISPERSION

```{r}

#Dispersion test and plot
bpu_dispr_loc <- vegan::betadisper(bpu_clr_dist_mat, phyloseq::sample_data(bpu_cpm_clr)$location)
bpu_dispr_loc

#Ordination Centroids and Dispersion Labeled: Aitchison Distance
bpu_ord_cen <- plot(bpu_dispr_loc, main = "", sub = "")
bpu_ord_cen

# extract data for ggplot
#ord_sites.long <- BiodiversityR::sites.long(ord_cen)
# boxplot
bpu_ord_box_loc <- boxplot(bpu_dispr_loc, main = "", xlab = "", xaxt='n')
tick <- seq_along(ord_box_loc$names)
axis(1, at = tick, labels = FALSE)
text(tick, par("usr")[2] - 0, bpu_ord_box_loc$names, srt = 45, xpd = TRUE)

# permutation test
permutest(bpu_dispr_loc)


```

##### ORDINATION

```{r}
# create sample_id variable
sample_data(bpu_cpm_clr)['sample_id'] <- row.names(sample_data(bpu_cpm_clr))

# set orination parameters
bpu_ord_all_clr <- phyloseq::ordinate(bpu_cpm_clr, "RDA")

#Plot scree plot
phyloseq::plot_scree(bpu_ord_all_clr) + 
  geom_bar(stat="identity", fill = "blue") +
  labs(x = "\nAxis", y = "Proportion of Variance\n")

#Scale axes and plot ordination
clr1 <- bpu_ord_all_clr$CA$eig[1] / sum(bpu_ord_all_clr$CA$eig)
clr2 <- bpu_ord_all_clr$CA$eig[2] / sum(bpu_ord_all_clr$CA$eig)

# plot ordination
bpu_ord_all_clr_loc_pca <- phyloseq::plot_ordination(bpu_cpm_clr, bpu_ord_all_clr, type = "samples", color="location")

# aesthetics of ordination plot
bpu_ord_all_clr_loc_pca <- bpu_ord_all_clr_loc_pca + 
  geom_point(size = 2) +
  coord_fixed(clr2 / clr1) +
  stat_ellipse(aes(group = location), linetype = 2) +
  labs(color = "Sample Location") +
  geom_text_repel(label = sample_data(bpu_cpm_clr)$sample_id, show.legend = FALSE) +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_text(face = "bold", size = 13),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12))

# print ordination plot
bpu_pca_allsam_loc <- print(bpu_ord_all_clr_loc_pca)

bpu_pca_allsam_loc

ggsave("output/plots/bouru-jessica-betadiv.pdf", bpu_pca_allsam_loc, dpi = 600)
```

### Alpha Div

#### Obs. Richness vs. Total Reads
```{r}
ggplot(data = data.frame("total_reads" =  phyloseq::sample_sums(bpu),
                         "observed" = phyloseq::estimate_richness(bpu, measures = "Observed")[, 1]),
       aes(x = total_reads, y = observed)) +
  geom_point() +
  geom_smooth(method="lm", se = FALSE) +
  labs(x = "\nTotal Reads", y = "Observed Richness\n")
```

> Positive upward trend between Observed Richness and total number of reads for each sample, as expected.

#### Shannon Div

> Boxplot visualizations with wilcox. test for significant differences in Shannon Diversity Index measurements. 

##### Location
###### Boxplot
```{r}
bpu_all_boxplot_loc <- plot_richness(bpu_cpm, x = "location", measures = c("Shannon"))#, title = "Microbiome Alpha Diversity: Samples by NDFA Group")
bpu_all_boxplot_loc <- bpu_all_boxplot_loc + geom_boxplot(aes(fill = location)) +
  labs(fill="Sample Location") +
  xlab("Location") +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.y = element_text(size = 10))

comps.bpu <- make_pairs(sample_data(bpu_cpm)$location)
print(comps)

adiv_loc.bpu <- bpu_all_boxplot_loc + stat_compare_means(
  comparisons = comps.bpu,
  label = "p.signif",
  tip.length = 0.05,
  symnum.args = list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
    symbols = c("xxxx", "***", "**", "*", "ns")
  ),
  method = "wilcox.test")

adiv_loc.bpu

ggsave(filename = "./output/plots/bouru-jessica-alphadiv-plot.pdf", adiv_loc.bpu, device = "pdf", dpi = 600)

```

###### Points
```{r}
bpu_all_point_loc <- plot_richness(bpu_cpm, x = "location", measures = c("Shannon"))#, title = "Microbiome Alpha Diversity: Samples by NDFA Group")
bpu_all_point_loc <- bpu_all_point_loc + geom_point(aes(color = location)) +
  labs(color="Sample Location") +
  xlab("Location") +
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(face = "bold", size = 13),
        axis.text.y = element_text(size = 10))

comps.bpu <- make_pairs(sample_data(bpu_cpm)$location)
print(comps)

adiv_loc.bpu <- bpu_all_point_loc + stat_compare_means(
  comparisons = comps.bpu,
  label = "p.signif",
  tip.length = 0.05,
  symnum.args = list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
    symbols = c("xxxx", "***", "**", "*", "ns")
  ),
  method = "wilcox.test")

adiv_loc.bpu

ggsave(filename = "./output/plots/bouru-jessica-alphadiv-point-plot.pdf", adiv_loc.bpu, device = "pdf", dpi = 600)

```


## Taxonomic Composition
### By Location

```{r}
library(phyloseq)
library(ggplot2)
library(dplyr)
library(tidyr)
library(randomcoloR)

# Set seed for reproducibility
set.seed(123)

### DECIDE how many taxa you want to see
topN = 20 # top 20 taxa

# Function to create a unique color palette
create_palette <- function(topN) {
  distinctColorPalette(topN + 1)
}

# Function to generate plot data for a given taxonomic rank
generate_plot_data <- function(tax_rank, topN, physeq_obj) {
  # Create a data frame of top taxa
  topTaxa <- physeq_obj %>% 
    psmelt() %>% 
    group_by_at(vars(one_of(tax_rank))) %>%  # Dynamic grouping based on tax_rank
    summarise(Abundance = sum(Abundance)) %>%
    arrange(desc(Abundance)) %>%
    slice_max(order_by = Abundance, n = topN) %>%  # Correct placement of slice_max
    mutate(aggTaxo = factor(if_else(row_number() <= topN, as.character(.[[tax_rank]]), 'Others'))) %>%
    ungroup()  # Ensure the data is ungrouped before proceeding

  # Ensure the others_placeholder has the same structure as topTaxa
  others_placeholder <- tibble(aggTaxo = factor('Others'))
  if (ncol(topTaxa) > 1) {
    # Add the same columns as in topTaxa, filled with appropriate default values
    others_placeholder[setdiff(names(topTaxa), "aggTaxo")] <- NA
  }
  
  # Ensure that 'Others' category is included in topTaxa
  topTaxa <- rbind(topTaxa, others_placeholder)

  # Create plot data by joining with a placeholder for "Others" to include it in the plot even if it's not in the topTaxa
  plot_data <- physeq_obj %>%
    psmelt() %>%
    mutate(aggTaxo = if_else(as.character(.[[tax_rank]]) %in% topTaxa[[tax_rank]], as.character(.[[tax_rank]]), 'Others')) %>%
    group_by(location, aggTaxo) %>%
    summarise(Abundance = sum(Abundance), .groups = 'drop')  # Use .groups='drop' to avoid grouping issues

  # Calculate total abundance for each location
    total_abundance <- plot_data %>%
        group_by(location) %>%
        summarise(TotalAbundance = sum(Abundance), .groups = 'drop')

    # Join total abundance with plot_data
    plot_data <- left_join(plot_data, total_abundance, by = "location")

    # Calculate relative abundance
    plot_data <- plot_data %>%
        mutate(RelativeAbundance = round(Abundance / TotalAbundance, 4))

    # Create summary dataframe with normalized counts
    summary_counts <- plot_data %>%
        select(aggTaxo, location, Abundance) %>%
        spread(key = location, value = Abundance)

    # Create summary dataframe with relative abundances
    summary_relative_abundance <- plot_data %>%
        select(aggTaxo, location, RelativeAbundance) %>%
        spread(key = location, value = RelativeAbundance)

    # Replace NA with 0
    summary_counts[is.na(summary_counts)] <- 0
    summary_relative_abundance[is.na(summary_relative_abundance)] <- 0

    return(list(plot_data = plot_data, 
                topTaxa = topTaxa, 
                summary_counts = summary_counts, 
                summary_relative_abundance = summary_relative_abundance))
}

# Function to plot data
plot_abundance <- function(plot_data, topTaxa, tax_rank, mycolors) {
  ggplot(plot_data, aes(x = location, y = Abundance, fill = aggTaxo)) +
    geom_bar(stat = "identity", position = "fill") +
    labs(title = paste("Top", tax_rank, "by location"),
         fill = tax_rank, x = "Location", y = "Relative abundance") +
    scale_fill_manual(values = mycolors) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          legend.title = element_blank())
}

# Generate unique color palette for the top taxa
mycolors <- create_palette(topN)

# Loop through the desired taxonomic levels to generate plots and store summary dataframes
for(tax_rank in c("Phylum", "Family", "Genus", "Species2")) {
  # Generate plot data
  plot_info <- generate_plot_data(tax_rank, topN, bpu_cpm) # insert phyloseq object name in last position
  plot_data <- plot_info$plot_data
  topTaxa <- plot_info$topTaxa
  counts_df <- plot_info$summary_counts
  relab_df <- plot_info$summary_relative_abundance
  
  # Define file names for plot and summary CSV
  plot_file_name <- paste0("./output/plots/bpuru-jessica-relab_top_", tax_rank, "_location.pdf")
  summary_counts_file_name <- paste0("./output/tables/bpuru-jessica-relab_top_", tax_rank, "_counts.csv")
  summary_relab_file_name <- paste0("./output/tables/bpuru-jessica-relab_top_", tax_rank, "_relab.csv")

  # Plot and save the figure
  plot <- plot_abundance(plot_data, topTaxa, tax_rank, mycolors)
  print(plot)
  ggsave(plot_file_name, plot, width = 10, height = 8)

  # Write cpm count dataframe to CSV
  write.csv(counts_df, summary_counts_file_name, row.names = FALSE)
  
  # Write relative abundance dataframe to CSV
  write.csv(relab_df, summary_relab_file_name, row.names = FALSE)
}

```

```{r}
tax_rank(bpu_cpm)
```


```{r}

otu_mat <- otu_table(bpu_cpm)

if (class(otu_mat) == "dgCMatrix") {
  otu_mat <- as(otu_mat, "matrix")
}

otu_df <- as.data.frame(otu_mat)

#otu_df <- t(otu_df)

taxa <- tax_table(bpu_cpm)
if (!is.null(taxa)) {
  if (class(taxa) == "dgCMatrix") {
    taxa <- as(taxa, "matrix")
  }
  taxa_df <- as.data.frame(taxa)

  # Make sure that the rownames of taxa_df are the same as the column names of otu_df
  # This should be true if the data has not been manipulated since creation
  # If they are not, you should align them by ordering or other methods as needed.
  rownames(otu_df) <- taxa_df$genus # or whichever taxonomic rank you prefer, like taxa_df$Genus
}

# You can access it directly for analyses or write it to a CSV file
write.csv(otu_df, file = "otu_table.csv", row.names = TRUE)
```




 

